File: intersection_of_2_arrays.cpp
// Time Complexity : O(m+n)
// Space Complexity :  O(m+n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        
        int len1 = nums1.size();
        int len2 = nums2.size();
        int i;
        
        map<int, int> m1;
        map<int, int> m2;
        
        vector<int> ret_vect;
        
        for(i = 0; i < len1; i++)
        {
            if(m1.find(nums1[i]) == m1.end())
                m1.insert(pair<int, int>(nums1[i], 1));
            else
                m1[nums1[i]] += 1;
        }
        
        for(i = 0; i < len2; i++)
        {
            if(m2.find(nums2[i]) == m2.end())
                m2.insert(pair<int, int>(nums2[i], 1));
            else
                m2[nums2[i]] += 1;
        }
        
        map<int, int>::iterator itr;
        int t1, t2, t;
        
        for(itr = m1.begin(); itr != m1.end(); itr++)
        {
            if(m2.find(itr->first) != m2.end())
            {
                t1 = itr->second;
                t2 = m2[itr->first];
                
                t = min(t1, t2);
                
                for(i = 0; i < t; i++)
                    ret_vect.push_back(itr->first);
            }
            
        }
        
        return ret_vect;
    }
};

File: Intersection of 2 arrays.py
# Time complexity : O(m+n)
# Space complexity : O(n) to maintain the hash map

# The code ran on Leetcode

# Maintain a hash map of smaller array. Iterate over the larger array, check if it is already present in the hash map. Add it to the result if map[n] > 0

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) < len(nums2):
            nums1, nums2 = nums2, nums1
        res = []
        map = {}
        # Get the counter hash map
        for n in nums2:
            if n in map:
                map[n] += 1
            else:
                map[n] = 1

        for n in nums1:
            if n in map and map[n] > 0:
                res.append(n)
                map[n] -= 1
        return res
        
